### DISCUSSION TITLE
`using` declaration on any block

### DISCUSSION THREAD

**Discussion by maxtoroq** [dotnet/csharplang#6623] [Language Ideas]
Example:

``` csharp
class MyClass {
   using Foo;

   void MyMethod() {
      using Bar;

      // Foo and Bar visible

      using Baz;

      // Foo, Bar and Baz visible
   }
}
```

**Comment by DavidArno** [comment]
What would `using Foo;` actually end up as in any expanded, compile-time, code? Presumably `using Foo` is just shorthand for eg 

``` cs
private readonly using Foo SomeFoo = new Foo();
```

and thus you'd presumably want `Myclass.Finalize()` to be auto-implemented/extended at compile time to ensure `((IDisposable)Foo).Dispose();` was called when the `MyClass` instance was garbage collected?

**Comment by HaloFour** [comment]
@maxtoroq What exactly are `Foo`, `Bar` or `Baz`?  How does this translate to existing code?  It's difficult to tell because you provide very little context and your code sample seems very incomplete, but if I had to guess I would say that it's a dupe of dotnet/roslyn#181.

**Comment by alrz** [comment]
I think this ain't about disposables. It's more likely to be a duplicate of dotnet/roslyn#4491.

**Comment by HaloFour** [comment]
Ah, makes sense.  Context is useful.

**Comment by alrz** [comment]
Using this you can make yourself a scoped `Self` type (if you have multiple types in a file),

``` cs
class C
{
  using Self = C;
}
```

You can use an analyzer to make sure that type aliases named `Self` are identical to the enclosing type.

**Comment by Thaina** [comment]
I think  this would also enable my proposal

https://github.com/dotnet/roslyn/issues/16958
